File: app/src/main/java/com/divercity/android/data/networking/services/DataService.java
package com.divercity.android.data.networking.services;

import com.divercity.android.data.entity.base.DataArray;
import com.divercity.android.data.entity.company.createcompanybody.CreateCompanyBody;
import com.divercity.android.data.entity.company.response.CompanyResponse;
import com.divercity.android.data.entity.company.sizes.CompanySizeResponse;
import com.divercity.android.data.entity.industry.IndustryResponse;
import com.divercity.android.data.entity.interests.InterestsResponse;
import com.divercity.android.data.entity.location.LocationResponse;
import com.divercity.android.data.entity.major.MajorResponse;
import com.divercity.android.data.entity.occupationofinterests.OOIResponse;
import com.divercity.android.data.entity.school.SchoolResponse;
import com.divercity.android.data.entity.skills.SkillResponse;

import io.reactivex.Observable;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Query;

/**
 * Created by lucas on 09/09/2018.
 */

public interface DataService {

    @GET("jobs/job_employers")
    Observable<Response<DataArray<CompanyResponse>>> fetchCompanies(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("data/industries")
    Observable<Response<DataArray<IndustryResponse>>> fetchIndustries(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("schools")
    Observable<Response<DataArray<SchoolResponse>>> fetchSchools(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("data/student_majors")
    Observable<Response<DataArray<MajorResponse>>> fetchStudentMajors(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("data/company_sizes")
    Observable<Response<DataArray<CompanySizeResponse>>> fetchCompanySizes();

    @GET("data/cities")
    Observable<Response<DataArray<LocationResponse>>> fetchLocations(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("data/job_skills")
    Observable<Response<DataArray<SkillResponse>>> fetchSkills(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("data/occupation_of_interests")
    Observable<Response<DataArray<OOIResponse>>> fetchOccupationOfInterests(
            @Query("page[number]") int pageNumber,
            @Query("page[size]") int size,
            @Query("search_query") String query);

    @GET("interests")
    Observable<Response<DataArray<InterestsResponse>>> fetchInterests();

    @POST("job_employers")
    Observable<Response<Void>> createCompany(@Body CreateCompanyBody body);
}

File: app/src/main/java/com/divercity/android/data/networking/services/DataService.kt
package com.divercity.android.data.networking.services

import com.divercity.android.data.entity.base.DataArray
import com.divercity.android.data.entity.company.createcompanybody.CreateCompanyBody
import com.divercity.android.data.entity.company.response.CompanyResponse
import com.divercity.android.data.entity.company.sizes.CompanySizeResponse
import com.divercity.android.data.entity.degree.DegreeEntityResponse
import com.divercity.android.data.entity.industry.IndustryResponse
import com.divercity.android.data.entity.interests.InterestsResponse
import com.divercity.android.data.entity.location.LocationResponse
import com.divercity.android.data.entity.major.MajorEntityResponse
import com.divercity.android.data.entity.occupationofinterests.OOIResponse
import com.divercity.android.data.entity.photo.PhotoEntityResponse
import com.divercity.android.data.entity.recommendedjobsgoi.RecommendedJobsGOIResponse
import com.divercity.android.data.entity.school.SchoolEntityResponse
import com.divercity.android.data.entity.skills.SkillResponse

import io.reactivex.Observable
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

/**
 * Created by lucas on 09/09/2018.
 */

interface DataService {

    @GET("jobs/job_employers")
    fun fetchCompanies(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<CompanyResponse>>>

    @GET("recommenders/group_of_interests_jobs")
    fun fetchRecommendedJobsGOIs(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int
    ): Observable<RecommendedJobsGOIResponse>

    @GET("data/industries")
    fun fetchIndustries(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<IndustryResponse>>>

    @GET("schools")
    fun fetchSchools(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<SchoolEntityResponse>>>

    @GET("data/student_majors")
    fun fetchStudentMajors(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<MajorEntityResponse>>>

    @GET("data/company_sizes")
    fun fetchCompanySizes(): Observable<Response<DataArray<CompanySizeResponse>>>

    @GET("data/cities")
    fun fetchLocations(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<LocationResponse>>>

    @GET("data/job_skills")
    fun fetchSkills(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<SkillResponse>>>

    @GET("data/occupation_of_interests")
    fun fetchOccupationOfInterests(
        @Query("page[number]") pageNumber: Int,
        @Query("page[size]") size: Int,
        @Query("search_query") query: String?
    ): Observable<Response<DataArray<OOIResponse>>>

    @GET("interests")
    fun fetchInterests(): Observable<Response<DataArray<InterestsResponse>>>

    @GET("group_of_interests/suggest_group_photo")
    fun searchPhotos(
        @Query("query") query: String
    ): Observable<Response<DataArray<PhotoEntityResponse>>>

    @POST("job_employers")
    fun createCompany(@Body body: CreateCompanyBody): Observable<Response<Void>>

    @GET("data/degrees")
    fun fetchDegrees(): Observable<Response<DataArray<DegreeEntityResponse>>>
}
